src/eos-updater-poll.c

```diff
@@ -203,6 +203,192 @@ get_fetchers (SourcesConfig *config,
 }
 
 static void
+object_unref0 (gpointer obj)
+{
+  if (obj == NULL)
+    return;
+  g_object_unref (obj);
+}
+
+static GPtrArray *
+get_finders (SourcesConfig          *config,
+             GMainContext           *context,
+             OstreeRepoFinderAvahi **out_finder_avahi)
+{
+  g_autoptr(OstreeRepoFinderAvahi) finder_avahi = NULL;
+  g_autoptr(GPtrArray) finders = g_ptr_array_new_full (config->download_order->len,
+                                                       object_unref0);
+  g_autoptr(GError) local_error = NULL;
+  gsize i;
+
+  g_assert (config->download_order->len > 0);
+
+  for (i = 0; i < config->download_order->len; i++)
+    {
+      switch (g_array_index (config->download_order, EosUpdaterDownloadSource, i))
+        {
+        case EOS_UPDATER_DOWNLOAD_MAIN:
+          g_ptr_array_add (finders, ostree_repo_finder_config_new ());
+          break;
+
+        case EOS_UPDATER_DOWNLOAD_LAN:
+          finder_avahi = ostree_repo_finder_avahi_new (context);
+          g_ptr_array_add (finders, g_object_ref (finder_avahi));
+          break;
+
+        case EOS_UPDATER_DOWNLOAD_VOLUME:
+          /* TODO: How to make this one testable? */
+          g_ptr_array_add (finders, ostree_repo_finder_mount_new (NULL));
+          break;
+
+        default:
+          g_assert_not_reached ();
+        }
+    }
+
+  g_ptr_array_add (finders, NULL);  /* NULL terminator */
+
+  /* TODO: Stop this at some point; think of a better way to store it and
+   * control its lifecycle. */
+  if (finder_avahi != NULL)
+    ostree_repo_finder_avahi_start (OSTREE_REPO_FINDER_AVAHI (finder_avahi),
+                                    &local_error);
```

Comment:
Does this return a `gboolean`? If so, can we check the result of that instead of checking `local_error` below?

Reply:
It does not. I could change it to in upstream OSTree, but I don’t want to block this branch on getting another change into OSTree (even if it is trivial).

Reply:
Makes sense